### PR TITLE
Create devnet-mint.test.ts

### PR DESCRIPTION
**Pull request by shramee**
For Shard-Labs/starknet-hardhat-plugin#272
Pull Shard-Labs/starknet-hardhat-plugin#285

### CODE REVIEW

File: test/devnet-mint.test.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        address = accounts[0].address;
+    });
+
+    it("should contain balance", async () => {
+        const balanceResponse = await starknet.devnet.mint(address, 0, true);
+        balance = balanceResponse.new_balance;
+
+        expect(typeof balance).to.equal("number");
+    });
+    it("shuold mint a quarter eth", async () => {
```
Comment (FabijanC) [inline test/devnet-mint.test.ts:24]: typo

File: test/devnet-mint.test.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    before(async () => {
+        const accounts = await starknet.devnet.getPredeployedAccounts();
+        address = accounts[0].address;
+    });
+
+    it("should contain balance", async () => {
+        const balanceResponse = await starknet.devnet.mint(address, 0, true);
+        balance = balanceResponse.new_balance;
+
+        expect(typeof balance).to.equal("number");
```
Comment (FabijanC) [inline test/devnet-mint.test.ts:22]: Don't we also want to assert other response elements?

File: test/devnet-mint.test.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+describe("Devnet mint", function () {
+    this.timeout(TIMEOUT);
+    let balance = 0,
+        address = "";
+    const div = 1e16; // To round off large numbers
+    const diff = 25 * div;
+
+    before(async () => {
+        const accounts = await starknet.devnet.getPredeployedAccounts();
```
Comment (FabijanC) [inline test/devnet-mint.test.ts:14]: It doesn't have to be a predeployed account. The address being funded doesn't even have to be that of a deployed account (makes sense if you are deploying a new account - the address needs to be funded even though the account is not yet on the network).

Therefore a simple dummy address should work (and probably no need for the `before` block)

File: test/devnet-mint.test.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        address = accounts[0].address;
+    });
+
+    it("should contain balance", async () => {
+        const balanceResponse = await starknet.devnet.mint(address, 0, true);
```
Comment (FabijanC) [inline test/devnet-mint.test.ts:19]: Would it make sense to test both a lite and a non-lite (regular) minting?

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: test/devnet-mint.test.ts (hunk 3 above)
Comment (shramee) [reply test/devnet-mint.test.ts:14]: Aah, aye cool! Thanks!

File: test/devnet-mint.test.ts (hunk 4 above)
Comment (shramee) [reply test/devnet-mint.test.ts:19]: Yeap, tots would.

File: test/devnet-mint.test.ts (hunk 2 above)
Comment (shramee) [reply test/devnet-mint.test.ts:22]: Sure can.

File: test/devnet-mint.test.ts (hunk 1 above)
Comment (shramee) [reply test/devnet-mint.test.ts:24]: Woops 🙊

**Comment by shramee** [comment]
@FabijanC Ready for another look. Should be much nicer now!

File: test/devnet-mint.test.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    });
+
+    it("should tx_hash in separate transaction (not lite) mode", async () => {
+        const resp = await starknet.devnet.mint(address, 0, true);
+        const { new_balance, unit, tx_hash } = resp;
+        balance = new_balance;
+
+        expect(typeof new_balance).to.equal("number");
+        expect(unit).to.equal("wei");
+        expect(tx_hash).to.equal(null);
```
Comment (FabijanC) [inline test/devnet-mint.test.ts:30]: But why is this `null` if there is a transaction?

To assert that it's a hex string, you can do
```ts
expect(tx_hash.startsWith("0x")).to.be.true
```

File: test/devnet-mint.test.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import { expect } from "chai";
+import { starknet } from "hardhat";
+
+import { TIMEOUT } from "./constants";
+
+describe("Devnet mint", function () {
+    this.timeout(TIMEOUT);
+    let balance = 0;
+    const address = "0x22edb233f3bfb99d2e9a14d283794515d124ef94398e143906d379348693fb7";
```
Comment (FabijanC) [inline test/devnet-mint.test.ts:9]: just comment here that it's a randomly selected address

File: test/devnet-mint.test.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        balance = new_balance;
+
+        expect(typeof new_balance).to.equal("number");
+        expect(unit).to.equal("wei");
+        expect(tx_hash).to.equal(null);
+    });
+
+    it("should mint lite mode", async () => {
+        const resp = await starknet.devnet.mint(address, diff, true);
+        expect(Math.round(resp.new_balance / div - balance / div)).to.deep.equal(diff / div);
```
Comment (FabijanC) [inline test/devnet-mint.test.ts:35]: If I might suggest extracting this somewhat complex logic of checking. Although I'm not completely sure what's the benefit of it? Why not just expect strict equality of values? If you're worried about losing precision due to the nature of numbers in JS, you can use something that's reasonably small.

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: test/devnet-mint.test.ts (hunk 7 above)
Comment (shramee) [reply test/devnet-mint.test.ts:35]: True, back when I was using pre funded accounts it was an issue, not anymore. Removed.

File: test/devnet-mint.test.ts (hunk 5 above)
Comment (shramee) [reply test/devnet-mint.test.ts:30]: Aah, I forgot to switch it to `false` for lite, which made the tests pass. Fixed.

File: test/devnet-mint.test.ts [hunk 8]
```diff
@@ -21,23 +21,26 @@ describe("Devnet mint", function () {
...
         expect(unit).to.equal("wei");
-        expect(tx_hash).to.equal(null);
+        expect(typeof tx_hash).to.equal("string");
+        expect(tx_hash.indexOf("0x")).to.equal(0);
     });
 
     it("should mint lite mode", async () => {
         const resp = await starknet.devnet.mint(address, diff, true);
-        expect(Math.round(resp.new_balance / div - balance / div)).to.deep.equal(diff / div);
+        expect(Math.round(resp.new_balance)).to.deep.equal(diff + balance);
```
Comment (FabijanC) [inline test/devnet-mint.test.ts:37]: No need for rounding, right?

Comment (shramee) [reply test/devnet-mint.test.ts]: Aah, sheesh!
Done now.

**Comment by FabijanC** [comment]
Changing target branch to `plugin` (this is mentioned in the TODO list of the plugin-repo PR)

**Pull request merged**